File: packages/e2e-test/src/__tests__/toolRename.test.ts
import { Architecture, Platform } from "@dotfiles/core";

import { describe, expect, it } from "bun:test";
import fs from "node:fs";
import path from "node:path";
import { TestHarness } from "./helpers/TestHarness";

const ORIGINAL_TOOL_CONFIG =
  "import { defineTool } from '@dotfiles/cli';\n\nexport default defineTool((install) =>\n  install().symlink('./config.txt', '~/.config/renameable-tool/config.txt')\n);\n";

describe("E2E: tool rename cleanup", () => {
  const harness = new TestHarness({
    testDir: import.meta.dir,
    configPath: "fixtures/tool-rename/config.ts",
    platform: Platform.MacOS,
    architecture: Architecture.Arm64,
  });

  const toolDir = path.join(import.meta.dir, "fixtures/tool-rename/tools/renameable-tool");
  const oldToolConfigPath = path.join(toolDir, "old-tool.tool.ts");
  const newToolConfigPath = path.join(toolDir, "new-tool.tool.ts");

  it("should stop reporting orphan cleanup after the renamed tool is cleaned once", async () => {
    await harness.clean();
    await fs.promises.rm(newToolConfigPath, { force: true });
    await fs.promises.writeFile(oldToolConfigPath, ORIGINAL_TOOL_CONFIG);

    try {
      const initialGenerateResult = await harness.generate();
      expect(initialGenerateResult.code).toBe(0);

      await fs.promises.rename(oldToolConfigPath, newToolConfigPath);

      const cleanupGenerateResult = await harness.generate();
      expect(cleanupGenerateResult.code).toBe(0);

      const repeatedGenerateResult = await harness.generate();
      expect(repeatedGenerateResult.code).toBe(0);
      expect(repeatedGenerateResult.stdout).toBe(
        ["WARN\tPlatform overridden to: macos", "WARN\tArch overridden to: arm64", "INFO\tDONE"].join("\n"),
      );
    } finally {
      await fs.promises.rm(newToolConfigPath, { force: true });
      await fs.promises.writeFile(oldToolConfigPath, ORIGINAL_TOOL_CONFIG);
      await harness.clean();
    }
  });
});

File: tests/e2e/tool_rename_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const originalToolConfig = `import { defineTool } from "@dotfiles/cli";

export default defineTool((install) => install().symlink("./config.txt", "~/.config/renameable-tool/config.txt"));
`

func TestE2EToolRename(t *testing.T) {
	t.Parallel()

	h := NewTestHarness(t, HarnessOptions{
		ConfigPath: "config.ts",
	})

	h.CopyFixture("tool-rename")

	// Ensure the source file config.txt is placed in the root of TempDir where the CLI resolves relative paths
	err := os.WriteFile(filepath.Join(h.TempDir, "config.txt"), []byte("config content\n"), 0644)
	if err != nil {
		t.Fatalf("failed to write config.txt source file: %v", err)
	}

	toolDir := filepath.Join(h.TempDir, "tools", "renameable-tool")
	oldToolConfigPath := filepath.Join(toolDir, "old-tool.tool.ts")
	newToolConfigPath := filepath.Join(toolDir, "new-tool.tool.ts")

	// Ensure fresh state
	_ = os.Remove(newToolConfigPath)
	err = os.WriteFile(oldToolConfigPath, []byte(originalToolConfig), 0644)
	if err != nil {
		t.Fatalf("failed to write original tool config: %v", err)
	}

	// First generate
	stdout, stderr, exitCode, err := h.Generate()
	if err != nil || exitCode != 0 {
		t.Fatalf("initial generate failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
	}

	// Rename the tool config
	err = os.Rename(oldToolConfigPath, newToolConfigPath)
	if err != nil {
		t.Fatalf("failed to rename tool config: %v", err)
	}

	// Second generate (triggers orphan/stale tool cleanup)
	stdout, stderr, exitCode, err = h.Generate()
	if err != nil || exitCode != 0 {
		t.Fatalf("generate after rename failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
	}

	// Runs generate a third time and checks it stops reporting orphan cleanup
	stdout, stderr, exitCode, err = h.Generate()
	if err != nil || exitCode != 0 {
		t.Fatalf("repeated generate failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
	}

	if strings.Contains(stdout, "Removing stale symlink") {
		t.Errorf("expected subsequent generate to skip orphan cleanup logs, but got:\n%s", stdout)
	}
}
